File: Best Time to Buy and Sell Stock.py
class Solution:
    """
    @param prices: Given an integer array
    @return: Maximum profit
    """
    def maxProfit(self, prices):
        # write your code here
        if not prices:
            return 0
        dp = []
        dp.append(0)
        minPrice = prices[0]
        for i in range(1,len(prices)):
            dp.append(max(dp[i - 1], prices[i] - minPrice))
            if prices[i] < minPrice:
                minPrice = prices[i]
        return dp[-1]


if __name__ == '__main__':
    a = Solution()
    print a.maxProfit([3,2,3,1,2])

File: java edition/bestTimeToBuyAndSellStock.java
public class bestTimeToBuyAndSellStock{
    public static void main(String[] args){
        int[] test = new int[]{3, 2, 3, 1, 2};
        System.out.println(maxProfit(test));
    }

    public static int maxProfit(int[] prices){
        if(prices == null || prices.length == 0){
            return 0;
        }

        int minSum = Integer.MAX_VALUE;
        int sum = 0;
        int max = Integer.MIN_VALUE;

        for(int i : prices){
            sum += i;
            minSum = (minSum > sum)? sum : minSum;
            max = ((i - minSum) > max )? (i - minSum): max;
        }
        return max;
    }
}
